### ISSUE TITLE
[proposal] "no-await-in-finally" -- detect `await` expressions that caller won't wait for

### ISSUE THREAD

**Issue by ide** [eslint/eslint#9099]
**Please describe what the rule should do:**
This rule checks if there are `await` expressions within `finally` clauses that the caller of the async function won't actually wait for.

```js
async function shuffleSomeFilesAsync() {
  try {
    const result = await someFileSystemStuffAsync();
    return result;  // early return
    // same issue exists if there's an uncaught error, it's basically returning early too
  } finally {
    await cleanupTempFilesAsync();
  }
}

async function() {
  const result = await shuffleSomeFilesAsync();
  // This below may fail! Awaiting on the method call above does not await upon
  // its finally clause because of the early return!
  assert(tempFilesAreCleanedUp());
}
```

See the code examples below.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// WARN
async function() {
  try {
    return; 
  } finally {
    await somethingAsync();
  }
}

// WARN
async function() {
  try {
  } catch (e) {
    throw e;
  } finally {
    await somethingAsync();
  }
}

// WARN
async function() {
  try {
  } catch (e) {
    return;
  } finally {
    await somethingAsync();
  }
}

// WARN
async function() {
  try {
    somethingThatCouldThrow(); 
  } finally {
    await somethingAsync();
  }
}

// OK (special case)
async function() {
  try {
    somethingThatCouldThrow(); 
  } catch (e) {
    // Special case: all errors caught, empty catch clause, no code here that could throw
  } finally {
    await somethingAsync();
  }
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule applies to plain JavaScript and catches an issue I bet a lot of people aren't aware of, is fairly generic, should be atomic & unique, and is independent of any specific library or framework.

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. I'm a bit confused because I'm not sure the issue you're describing actually exists.

I tested it with the following code:

```js
// returns a promise that fulfills after a given number of milliseconds
const delay = ms => new Promise(resolve => setTimeout(resolve, ms))

async function foo() {
  try {
    const result = await delay(1000);
    return result;
  } finally {
    await delay(1000);
    console.log('finally handler finished');
  }
}

foo().then(() => console.log('promise from async function fulfilled'));
```

If it was true that the caller of the async function doesn't wait for `await` expressions in the `finally` clause, then I would expect "promise from async function fulfilled" to be logged after 1 second, and then "finally handler finished" to be logged after 2 seconds.

However, when I actually run the code (in Chrome 60), "finally handler finished" is logged after 2 seconds, and then "promise from async function fulfilled" is logged immediately afterwards. So it seems like `await` expressions in the `finally` clause do, in fact, cause the resulting promise to be delayed.

**Comment by ide** [comment]
@not-an-aardvark You're totally right. I tested this in Node 8.3.0 and it has the behavior you described, but in Node 7 (7.7.1) it has the opposite behavior! Thanks for the extra pair of eyes, closing this out.
